### ISSUE TITLE
Further performance improvements.

### ISSUE THREAD

**Issue by sageserpent-open** [sageserpent-open/kineticMerge#35]
This follows on from #34, trying to win some more performance improvement.

The first thing to investigate is whether the subsumption and / or overlap tests for sections that form part of a potential match can be carried out quicker using an alternative implementation from the existing fingertree one - there are other data structures for this, such as interval trees and segment trees.

**Comment by sageserpent-open** [comment]
Another idea - further parallelization of fingerprinting. At time of writing (`main` at: f3b353220f1fad04064b4dc99b2d16210e72b71e), fingerprinting is executed in parallel across files

This could be extended so that for large files, fingerprinting takes place in parallel swathes. Although this means that each swathe needs to be primed beforehand (which is optimised away when the entire file is fingerprinted in one swathe), for large files on hopes that being able to parallelize outweighs this overhead.

**Comment by sageserpent-open** [comment]
Another idea - can the fingerprinting done by `RollingHash` use plain `Long` arithmetic as well as `BigInt`. I wonder whether this could be used either when looking for fingerprints for a small match window size, or perhaps as a sloppy approximation if the target files to be compared are all large (we have to make sure that fingerprints being compared across files are all computed in the same way to infer potential matches - well actually, we don't, but we may end up with a lot of spurious potential matches. Hmm...).

Worth doing some spikes on this idea...

**Comment by sageserpent-open** [comment]
Another idea - tiny matches (digraphs, trigraphs etc) can turn up ambiguously all over the place by coincidence and are thus useless for determining code motion. Think of:
```scala
       = {
              val
```

in a Scala 2 program, for example. That turns up at most function / method implementations.

Rather than waste time churning through all these useless matches umpteen times in the closing phase of finding matches where the match size is small, how about doing an exploratory search upfront, looking for (and then weeding out) such ambiguous matches. The results are then cached for the given match size and can be used during the closing phases of matching, avoiding overwhelming the match machinery with the ambiguous matches.

While this is potentially inaccurate in terms of forcing the match size guess back upwards to find an optimal match, I don't think it will yield incorrect results (other than the absence of the useless matches).

It could be packaged up as an automated minimal match-size option that does let smaller fry through if they don't make to much noise.

**Comment by sageserpent-open** [comment]
Another idea concerning match sizes - if we have a very large file and a bunch of small ones, finding smaller matches can get bogged down in constantly having to scan through the large file.

Given that the matching implementation suppresses useless matches that overlap within an existing larger optimal match, then we'd expect that at some point, unless there genuinely is a small match lurking in that large file, the search for matches in the large file will peter out. Once we've seen that, keeping track of the smallest match size seen so far for each file means that we can exclude the large file from all further searches (as that size won't be revisited).

That should save some churn in the matching endgame.

**Comment by sageserpent-open** [comment]
Example merge to glean some rough time measurements for a real-world merge:

Merge from  70b8c24381edcd5e576e6191e1cdb31bfaf87be3 into 2315cc54469cae72096c5a80c0ebf7c30ed0e737.

```bash
git checkout 2315cc54469cae72096c5a80c0ebf7c30ed0e737
kinetic-merge --no-commit 70b8c24381edcd5e576e6191e1cdb31bfaf87be3
```

4G heap allocation by default.

**Comment by sageserpent-open** [comment]
Taking a sample over match sizes from roughly 1500 to 280 (in the first 2 minutes or so):
![Screenshot 2024-12-10 at 09 59 52](https://github.com/user-attachments/assets/94fd11e7-8a4d-4837-bb27-c559da314e76)

![Screenshot 2024-12-10 at 10 00 55](https://github.com/user-attachments/assets/5404cee8-0f3e-41a7-96bb-433f8455507c)

Taking a sample over match sizes from roughly 1500 to 37 (in the first 15 minutes or so):

![Screenshot 2024-12-10 at 10 02 10](https://github.com/user-attachments/assets/d1bd5ed6-2707-4f70-aa6d-5af2541f03b2)
![Screenshot 2024-12-10 at 10 03 42](https://github.com/user-attachments/assets/9e9cfca7-8e25-4018-8fb5-1e9290de84bc)

Taking a sample over match sizes from roughly 34 downwards  (after 16 minutes or so):
![Screenshot 2024-12-10 at 10 04 25](https://github.com/user-attachments/assets/5d034af7-04a6-48a1-8383-a105bb1e57f6)
![Screenshot 2024-12-10 at 10 07 00](https://github.com/user-attachments/assets/f60a83d6-92dd-494a-8687-4b442211779d)

Taking a sample over match size 11 downwards  (after 45 minutes or so):
![Screenshot 2024-12-10 at 10 09 09](https://github.com/user-attachments/assets/8fe73404-6ddb-4030-91ab-3c6b60635cb4)
![Screenshot 2024-12-10 at 10 10 49](https://github.com/user-attachments/assets/bb49f03c-695b-46e7-a344-768ececa609c)

**Comment by sageserpent-open** [comment]
So a hasty conclusion would be to optimise the ordering *and* the fingertree.

Stepping back, it is clear that the majority of run time was spent in a later phase of behaviour at match size < 34, where the fingertree dominates. 

This is almost certainly down to the prevelance of small matches, perhaps causing ambiguous matches, or perhaps just lots of overlapping matches, most of which will be suppressed by larger matches anyway - but we need the fingertree to do this suppression.

Instead of just plunging into optimising the fingertree (or providing a better alternative), perhaps the ideas above could ameliorate the endgame?

Let's do this first...

**Comment by sageserpent-open** [comment]
Aside - is it best for `matchingFingerprintsAcrossSides` to work up through the potential match keys in order? This is efficient in that a single linear pass is made through each side's potential match keys, but might it be cheaper overall to work in terms of sets of keys, looking for the intersection of all three sides' sets to yield all-sides matches, and the disjunctions of pairwise intersection with the three-way intersection to look for pairwise matches?

**Comment by sageserpent-open** [comment]
Tried out the dead-zone idea where files are excluded from match detection when the binary chop moves its lower bound upwards in 91b023893aa08688e0cdeba6788161296a9491e6.

All tests pass, good.

The good news is that the application terminates after around 20-25 minutes (can't remember exactly, but is was a lot quicker).

The bad news is that instead of a merge, it bombs out when it constructs the filler sections after matching is complete, detecting overlapping sections of size 9.

Therefore a larger, optimal match was missed.

Examination of the output of `CodeMotionAnalysisTest.matchingSectionsAreFound` does not show any difference in the rejected test cases (which are due to overlaps), so far there's only the manual testing that exposes this.

My current theory is that the implementation is doing the right thing, but the strategy is slightly flawed - it is possible for a small match that is blocked by a larger *pairwise* match to be an overlapping part of an all-sides match intermediate in size between the pairwise match and the blocked match. That intermediate match *would* be permitted by virtue of eating into the larger pairwise match, so the effect is to skip over some optimal all-sides matches.

I'm surprised I haven't seen automated test failures for this, though...

**Comment by sageserpent-open** [comment]
The problem above is more subtle - it seems to be connected with the setting of `minimum-ambiguous-match-size`.

I suspect this is due to small ambiguous matches being completely rejected when they come within the threshold, thus fooling the path inclusion implementation into thinking it has encountered a dead-zone for some path.

Trying:

```bash
kinetic-merge --minimum-ambiguous-match-size=5 --no-commit 70b8c24381edcd5e576e6191e1cdb31bfaf87be3
```
 yielded a different failure where the overlapping sections were of size 6. Note that this ran in a whopping 53 minutes, almost certainly due to having to slog through loads more small ambiguous matches - the memory usage also went up at the end to around 2.7 Gb.

Let's try using an all-inclusive approach when we get down to the threshold...

**Comment by sageserpent-open** [comment]
This is implemented in abc892fedefe1f845952d8f4c48e427272cdafeb.

The example merge takes 35 minutes, so there is a modest improvement.

Unfortunately, in order to enjoy the benefits of path inclusion, it is necessary to avoid trimming ambiguous matches - so `minimum-ambiguous-match-size` is left at its default of 10. That is low enough in this example for lots of ambiguous matches to be generated, which thrashes the fingertree implementation.

**Comment by sageserpent-open** [comment]
Finessed this approach in 850d1885301d26d439167563efb044ba0d91e291, but still taking 35 minutes for the example merge.

**Comment by sageserpent-open** [comment]
Intriguingly, applying this patch:

```text
Index: src/main/scala/com/sageserpent/kineticmerge/core/CodeMotionAnalysis.scala
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/scala/com/sageserpent/kineticmerge/core/CodeMotionAnalysis.scala b/src/main/scala/com/sageserpent/kineticmerge/core/CodeMotionAnalysis.scala
--- a/src/main/scala/com/sageserpent/kineticmerge/core/CodeMotionAnalysis.scala	(revision 850d1885301d26d439167563efb044ba0d91e291)
+++ b/src/main/scala/com/sageserpent/kineticmerge/core/CodeMotionAnalysis.scala	(date 1733938953936)
@@ -1007,15 +1007,15 @@
               // Synchronised the fingerprints across all three sides...
               val potentialMatchesForSynchronisedFingerprint =
                 val baseSectionsThatDoNotOverlap = LazyList
-                  .from(baseSectionsByFingerprint.get(baseHead))
+                  .from(baseSectionsByFingerprint.get(baseHead)).take(10)
                   .filterNot(overlapsOrIsSubsumedByOnBase)
 
                 val leftSectionsThatDoNotOverlap = LazyList
-                  .from(leftSectionsByFingerprint.get(leftHead))
+                  .from(leftSectionsByFingerprint.get(leftHead)).take(10)
                   .filterNot(overlapsOrIsSubsumedByOnLeft)
 
                 val rightSectionsThatDoNotOverlap = LazyList
-                  .from(rightSectionsByFingerprint.get(rightHead))
+                  .from(rightSectionsByFingerprint.get(rightHead)).take(10)
                   .filterNot(overlapsOrIsSubsumedByOnRight)
 
                 for
@@ -1085,11 +1085,11 @@
               // Synchronised the fingerprints between the base and left...
               val potentialMatchesForSynchronisedFingerprint =
                 val baseSectionsThatDoNotOverlap = LazyList
-                  .from(baseSectionsByFingerprint.get(baseHead))
+                  .from(baseSectionsByFingerprint.get(baseHead)).take(10)
                   .filterNot(overlapsOrIsSubsumedByOnBase)
 
                 val leftSectionsThatDoNotOverlap = LazyList
-                  .from(leftSectionsByFingerprint.get(leftHead))
+                  .from(leftSectionsByFingerprint.get(leftHead)).take(10)
                   .filterNot(overlapsOrIsSubsumedByOnLeft)
 
                 for
@@ -1152,11 +1152,11 @@
               // Synchronised the fingerprints between the base and right...
               val potentialMatchesForSynchronisedFingerprint =
                 val baseSectionsThatDoNotOverlap = LazyList
-                  .from(baseSectionsByFingerprint.get(baseHead))
+                  .from(baseSectionsByFingerprint.get(baseHead)).take(10)
                   .filterNot(overlapsOrIsSubsumedByOnBase)
 
                 val rightSectionsThatDoNotOverlap = LazyList
-                  .from(rightSectionsByFingerprint.get(rightHead))
+                  .from(rightSectionsByFingerprint.get(rightHead)).take(10)
                   .filterNot(overlapsOrIsSubsumedByOnRight)
 
                 for
@@ -1219,11 +1219,11 @@
               // Synchronised the fingerprints between the left and right...
               val potentialMatchesForSynchronisedFingerprint =
                 val leftSectionsThatDoNotOverlap = LazyList
-                  .from(leftSectionsByFingerprint.get(leftHead))
+                  .from(leftSectionsByFingerprint.get(leftHead)).take(10)
                   .filterNot(overlapsOrIsSubsumedByOnLeft)
 
                 val rightSectionsThatDoNotOverlap = LazyList
-                  .from(rightSectionsByFingerprint.get(rightHead))
+                  .from(rightSectionsByFingerprint.get(rightHead)).take(10)
                   .filterNot(overlapsOrIsSubsumedByOnRight)
 
                 for

```
on top of commit 850d188 brings the exection time down to 19 minutes for the example merge.

This is welcome, but puzzling, because the sections are filtered as a lazy list and then checked to see if there is ambiguity - so my thoughts are that it should make little practical difference. However, it clearly does...

**Comment by sageserpent-open** [comment]
Applying the same patch on top of the 'original' commit of 34a0ff8b1d77a42146f193d264b1df80dd8b6238 results in an execution time of 23 minutes. Puzzling it may be, but it is clearly worth looking at...

**Comment by sageserpent-open** [comment]
Commit 539f59fa4cdcf58b653733546cd1a30382104a1c addresses this by using a hardwired maximum number of ambiguous matches for a given piece of matched content (after lazily vetting them for overlap with or subsumption of smaller sections).

Now, this technique is somewhat unsafe, because it can drop overlapping ambiguous matches encountered when searching for an optimal match of a run of the same token being repeated - this was observed in `CodeMotionAnalysisExtensionTest.codeMotionAmbiguousWithAPreservation` when the maximum number cutoff was set to 120 matches for the same content.

Setting the cutoff to 200 sorts that out. That test is rather pathological in that it deliberately sets up such runs of the same token being repeated - unless folk are trying to merge text such as:

```text
Wheeeeeeeeeeeeeeeeeeeeeeeeeeeee!Oooooooooooooooooooooh!Aaaaaaaaaaaaaaaaaaaaah.Eeeeeeeeeh?
```

or perhaps:

```java
class ThereAreManyLevelsToMe {
    void deepAndMeaningless() {

       {
           // Fenced off code block.

          class littleHelper{
               bool questionable(){
                     class cogInTheMachine{
                           {
                                // Initialiser block.

                               {
                                   // Another fenced-off block, etc....
                               }
                           }
                     }
               }
          }
       }
    }
}
```

I'm happy to regard both as being sufficiently esoteric to not warrant any bother.

Having said that, perhaps a command-line option to set the cutoff would be handy, as would logging of when the cutoff is exceeded and with what content...

Anyway, looking at a run time of 22 minutes for a rather ham-fisted attempt at this and 20 minutes for a more finessed form. In it goes...

**Comment by sageserpent-open** [comment]
Commit 22a979f683230d3e7bae60ea1ad84f7d5367e3c9 introduces the command-line flag `ambiguous-matches-threshold`.

Setting it to 200 yields this hot-spot report:
<img width="1743" alt="Screenshot 2024-12-12 at 12 28 34" src="https://github.com/user-attachments/assets/26fe71af-3038-4039-9765-689d786b9fc8" />

The ordering overhead has almost caught up with the fingertree overhead, as the latter is now diminished.

Setting it to 20 yields this:
<img width="1748" alt="Screenshot 2024-12-12 at 12 31 01" src="https://github.com/user-attachments/assets/129b8f30-7527-439b-9627-afad442ebe29" />

The ordering overhead is out in the front with a 2:1 ratio with the fingertree.

Execution time with a setting of 20 is down to 15 minutes, so we're on the right track.

**Comment by sageserpent-open** [comment]
Commit 42d2deb46c5cef6945b579ccaff00c086e5cbaa8 merges some work done on optimising the ordering, this brings the example's execution time down to 11 minutes, 11 minutes.

Hot spot report:
<img width="1726" alt="Screenshot 2024-12-12 at 13 42 08" src="https://github.com/user-attachments/assets/2bcb502a-1568-4e82-a059-aef2df47a5e4" />

Fingertree has moved back into the lead as ordering has diminished...

**Comment by sageserpent-open** [comment]
Commit 3091a0e13997e3d9df090157f177b575936cd3c6 has the work down in path inclusions merged in, plus a refresh from `main`.

The example's execution time is 10 minutes, so no real benefit there.

Hot spot report:
<img width="1739" alt="Screenshot 2024-12-13 at 13 26 37" src="https://github.com/user-attachments/assets/21809ad8-b965-4e47-b5b7-b8114b4a3b33" />

Again, fingertree is back to being the leading offender, but the ordering is right on its heels...

**Comment by sageserpent-open** [comment]
A closer look reveals that when the lower-level calls are included, the self-time is split more or less evenly between ordering and fingerprinting on the one hand and fingertree operations on the other.

So it is up for grabs as to which one to do first...

**Comment by sageserpent-open** [comment]
Ideas the haven't really yielded much benefit, but may be useful later...

Use a lazy field in `PotentialMatchKey` to avoid having to recompute content:

```text
Subject: [PATCH] WIP: use a lazy field in `PotentialMatchKey` to cache implied content, also use an immutable `SortedMultiDict`.
---
Index: src/main/scala/com/sageserpent/kineticmerge/core/CodeMotionAnalysis.scala
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/scala/com/sageserpent/kineticmerge/core/CodeMotionAnalysis.scala b/src/main/scala/com/sageserpent/kineticmerge/core/CodeMotionAnalysis.scala
--- a/src/main/scala/com/sageserpent/kineticmerge/core/CodeMotionAnalysis.scala	(revision 7c26c131ededa8eb7fc28022e9683055506c4b58)
+++ b/src/main/scala/com/sageserpent/kineticmerge/core/CodeMotionAnalysis.scala	(date 1734185104284)
@@ -15,7 +15,8 @@
 import scala.annotation.tailrec
 import scala.collection.immutable.{MultiDict, SortedMultiSet}
 import scala.collection.parallel.CollectionConverters.*
-import scala.collection.{SortedMultiDict, mutable}
+import scala.collection.mutable
+import scala.collection.immutable.SortedMultiDict
 import scala.util.Using
 
 trait CodeMotionAnalysis[Path, Element]:
@@ -1896,10 +1897,7 @@
         Order.by(_.fingerprint),
         // NOTE: need the pesky type ascription because `Order` is invariant on
         // its type parameter.
-        Order.by(
-          _.impliedContent.content
-            .take(tiebreakContentSamplingLimit): Seq[Element]
-        )
+        Order.by(_.content: Seq[Element])
       )
     end potentialMatchKeyOrder
 
@@ -1920,7 +1918,10 @@
     case class PotentialMatchKey(
         fingerprint: BigInt,
         impliedContent: Section[Element]
-    )
+    ):
+      lazy val content: IndexedSeq[Element] =
+        impliedContent.content.take(tiebreakContentSamplingLimit)
+    end PotentialMatchKey
 
     val matchesAndTheirSections =
       val withAllMatchesOfAtLeastTheSureFireWindowSize =
```

Convert the sets of `PotentialMatchKey` into lists for better decomposition scaling in `matchingFingerprintsAcrossSides`:

```text
Subject: [PATCH] WIP: convert the potential match key sets into lists for better decomposition.
---
Index: src/main/scala/com/sageserpent/kineticmerge/core/CodeMotionAnalysis.scala
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/scala/com/sageserpent/kineticmerge/core/CodeMotionAnalysis.scala b/src/main/scala/com/sageserpent/kineticmerge/core/CodeMotionAnalysis.scala
--- a/src/main/scala/com/sageserpent/kineticmerge/core/CodeMotionAnalysis.scala	(revision 7c26c131ededa8eb7fc28022e9683055506c4b58)
+++ b/src/main/scala/com/sageserpent/kineticmerge/core/CodeMotionAnalysis.scala	(date 1734182667112)
@@ -1342,10 +1342,13 @@
           end match
         end matchingFingerprintsAcrossSides
 
+        // NOTE: convert the three key sets of potential match keys to list so
+        // that decomposition inside `matchingFingerprintsAcrossSides` is
+        // efficient.
         matchingFingerprintsAcrossSides(
-          baseSectionsByFingerprint.keySet,
-          leftSectionsByFingerprint.keySet,
-          rightSectionsByFingerprint.keySet,
+          baseSectionsByFingerprint.keySet.toList,
+          leftSectionsByFingerprint.keySet.toList,
+          rightSectionsByFingerprint.keySet.toList,
           matches = Set.empty,
           haveTrimmedMatches = false
         )

```

**Comment by sageserpent-open** [comment]
> 2\. An alternative is to stick with the current approach of scanning up through triples of potential match keys, but to exploit the ordering more effectively by aggressively dropping keys up the next target instead of dropping them one-by-one. This might yield the same if not more benefit than the first approach.

Patch:

```text
Subject: [PATCH] WIP: skip forward more aggressively through the potential match keys when trying to find a match.
---
Index: src/main/scala/com/sageserpent/kineticmerge/core/CodeMotionAnalysis.scala
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/scala/com/sageserpent/kineticmerge/core/CodeMotionAnalysis.scala b/src/main/scala/com/sageserpent/kineticmerge/core/CodeMotionAnalysis.scala
--- a/src/main/scala/com/sageserpent/kineticmerge/core/CodeMotionAnalysis.scala	(revision 7c26c131ededa8eb7fc28022e9683055506c4b58)
+++ b/src/main/scala/com/sageserpent/kineticmerge/core/CodeMotionAnalysis.scala	(date 1734260498804)
@@ -13,9 +13,9 @@
 
 import java.lang.Byte as JavaByte
 import scala.annotation.tailrec
-import scala.collection.immutable.{MultiDict, SortedMultiSet}
+import scala.collection.immutable.{MultiDict, SortedMultiDict, SortedMultiSet}
+import scala.collection.mutable
 import scala.collection.parallel.CollectionConverters.*
-import scala.collection.{SortedMultiDict, mutable}
 import scala.util.Using
 
 trait CodeMotionAnalysis[Path, Element]:
@@ -981,9 +981,9 @@
 
         @tailrec
         def matchingFingerprintsAcrossSides(
-            baseFingerprints: Iterable[PotentialMatchKey],
-            leftFingerprints: Iterable[PotentialMatchKey],
-            rightFingerprints: Iterable[PotentialMatchKey],
+            baseFingerprints: collection.SortedSet[PotentialMatchKey],
+            leftFingerprints: collection.SortedSet[PotentialMatchKey],
+            rightFingerprints: collection.SortedSet[PotentialMatchKey],
             matches: Set[GenericMatch],
             haveTrimmedMatches: Boolean
         ): MatchingResult =
@@ -1059,7 +1059,7 @@
               matchingFingerprintsAcrossSides(
                 baseFingerprints,
                 leftFingerprints,
-                rightFingerprints.tail,
+                rightFingerprints.rangeFrom(baseHead),
                 matches,
                 haveTrimmedMatches
               )
@@ -1117,7 +1117,7 @@
               // right, need to advance the left to resolve ...
               matchingFingerprintsAcrossSides(
                 baseFingerprints,
-                leftFingerprints.tail,
+                leftFingerprints.rangeFrom(baseHead),
                 rightFingerprints,
                 matches,
                 haveTrimmedMatches
@@ -1175,7 +1175,7 @@
               // Tentatively synchronised the fingerprints between the left and
               // right, need to advance the base to resolve ...
               matchingFingerprintsAcrossSides(
-                baseFingerprints.tail,
+                baseFingerprints.rangeFrom(leftHead),
                 leftFingerprints,
                 rightFingerprints,
                 matches,
@@ -1228,21 +1228,22 @@
               // minimum fingerprint to see if it can catch up and
               // synchronise...
 
-              val minimumFingerprint =
-                potentialMatchKeyOrder.min(
-                  baseHead,
-                  potentialMatchKeyOrder
-                    .min(leftHead, rightHead)
-                )
+              val baseLeftMinimum =
+                potentialMatchKeyOrder.min(baseHead, leftHead)
+              val baseRightMinimum =
+                potentialMatchKeyOrder.min(baseHead, rightHead)
+              val leftRightMinimum =
+                potentialMatchKeyOrder.min(leftHead, rightHead)
 
               // NOTE: just use `==` as we have already looked inside the
               // `PotentialMatchKey` instances - we just want to know which one
               // was the minimum.
-              if leftHead == minimumFingerprint
+              if baseLeftMinimum == leftRightMinimum
               then
+                // The left is the minimum.
                 matchingFingerprintsAcrossSides(
                   baseFingerprints,
-                  leftFingerprints.tail,
+                  leftFingerprints.rangeFrom(baseRightMinimum),
                   rightFingerprints,
                   matches,
                   haveTrimmedMatches
@@ -1250,18 +1251,20 @@
               // NOTE: just use `==` as we have already looked inside the
               // `PotentialMatchKey` instances - we just want to know which one
               // was the minimum.
-              else if rightHead == minimumFingerprint
+              else if baseRightMinimum == leftRightMinimum
               then
+                // The right is the minimum.
                 matchingFingerprintsAcrossSides(
                   baseFingerprints,
                   leftFingerprints,
-                  rightFingerprints.tail,
+                  rightFingerprints.rangeFrom(baseLeftMinimum),
                   matches,
                   haveTrimmedMatches
                 )
               else
+                // The base is the minimum.
                 matchingFingerprintsAcrossSides(
-                  baseFingerprints.tail,
+                  baseFingerprints.rangeFrom(leftRightMinimum),
                   leftFingerprints,
                   rightFingerprints,
                   matches,
@@ -1276,7 +1279,7 @@
               if potentialMatchKeyOrder.lt(baseHead, leftHead)
               then
                 matchingFingerprintsAcrossSides(
-                  baseFingerprints.tail,
+                  baseFingerprints.rangeFrom(leftHead),
                   leftFingerprints,
                   rightFingerprints,
                   matches,
@@ -1285,7 +1288,7 @@
               else
                 matchingFingerprintsAcrossSides(
                   baseFingerprints,
-                  leftFingerprints.tail,
+                  leftFingerprints.rangeFrom(baseHead),
                   rightFingerprints,
                   matches,
                   haveTrimmedMatches
@@ -1298,7 +1301,7 @@
               if potentialMatchKeyOrder.lt(baseHead, rightHead)
               then
                 matchingFingerprintsAcrossSides(
-                  baseFingerprints.tail,
+                  baseFingerprints.rangeFrom(rightHead),
                   leftFingerprints,
                   rightFingerprints,
                   matches,
@@ -1308,7 +1311,7 @@
                 matchingFingerprintsAcrossSides(
                   baseFingerprints,
                   leftFingerprints,
-                  rightFingerprints.tail,
+                  rightFingerprints.rangeFrom(baseHead),
                   matches,
                   haveTrimmedMatches
                 )
@@ -1321,7 +1324,7 @@
               then
                 matchingFingerprintsAcrossSides(
                   baseFingerprints,
-                  leftFingerprints.tail,
+                  leftFingerprints.rangeFrom(rightHead),
                   rightFingerprints,
                   matches,
                   haveTrimmedMatches
@@ -1330,7 +1333,7 @@
                 matchingFingerprintsAcrossSides(
                   baseFingerprints,
                   leftFingerprints,
-                  rightFingerprints.tail,
+                  rightFingerprints.rangeFrom(leftHead),
                   matches,
                   haveTrimmedMatches
                 )
@@ -1342,10 +1345,16 @@
           end match
         end matchingFingerprintsAcrossSides
 
+        // NASTY HACK: have to work around `SortedMultiDict` not exposing the
+        // more refined type of its key sets, it simply inherits the type
+        // signature from `MultiDict`.
         matchingFingerprintsAcrossSides(
-          baseSectionsByFingerprint.keySet,
-          leftSectionsByFingerprint.keySet,
-          rightSectionsByFingerprint.keySet,
+          baseSectionsByFingerprint.keySet
+            .asInstanceOf[collection.SortedSet[PotentialMatchKey]],
+          leftSectionsByFingerprint.keySet
+            .asInstanceOf[collection.SortedSet[PotentialMatchKey]],
+          rightSectionsByFingerprint.keySet
+            .asInstanceOf[collection.SortedSet[PotentialMatchKey]],
           matches = Set.empty,
           haveTrimmedMatches = false
         )
```

This didn't much benefit, coming in at 11 minutes. There is a caveat that in order to skip forwards aggessively to move up to the next lowest potential match key, the patch relies on the keys being held in *sorted sets*. That is fine in itself, only the calls to `rangeFrom` rebuild a new sorted set, and that involves hitting the ordering some more.

Another approach is to convert the sets to indexed sequence, using a binary chop (and the implied sorted order of the sequences) to decompose them. The problem is, this means that each set has to be fully traversed by the conversion - while that doesn't hammer the ordering, it feels like unnessary overhead.

Perhaps this could be finessed by using a kind of sorted-set view that doesn't actually rebuild the set?

**Comment by sageserpent-open** [comment]
Had another go at the second option discussed aboce, this time avoiding decomposition of the key sets to prevent overhead; leaving them as sorted sets for efficient lookup.

Patch:

```text
Subject: [PATCH] WIP: skip forward more aggressively through the potential match keys when trying to find a match - this time, don't decompose the key sets.
---
Index: src/main/scala/com/sageserpent/kineticmerge/core/CodeMotionAnalysis.scala
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/scala/com/sageserpent/kineticmerge/core/CodeMotionAnalysis.scala b/src/main/scala/com/sageserpent/kineticmerge/core/CodeMotionAnalysis.scala
--- a/src/main/scala/com/sageserpent/kineticmerge/core/CodeMotionAnalysis.scala	(revision 7c26c131ededa8eb7fc28022e9683055506c4b58)
+++ b/src/main/scala/com/sageserpent/kineticmerge/core/CodeMotionAnalysis.scala	(date 1734268989044)
@@ -13,9 +13,9 @@
 
 import java.lang.Byte as JavaByte
 import scala.annotation.tailrec
-import scala.collection.immutable.{MultiDict, SortedMultiSet}
+import scala.collection.immutable.{MultiDict, SortedMultiDict, SortedMultiSet}
+import scala.collection.mutable
 import scala.collection.parallel.CollectionConverters.*
-import scala.collection.{SortedMultiDict, mutable}
 import scala.util.Using
 
 trait CodeMotionAnalysis[Path, Element]:
@@ -943,7 +943,7 @@
           accumulatingResults
         end fingerprintStartIndices
 
-        def fingerprintSections(
+        def sectionsByPotentialMatchKey(
             sources: Sources[Path, Element],
             pathIsIncluded: Path => Boolean
         ): SortedMultiDict[PotentialMatchKey, Section[Element]] =
@@ -970,32 +970,50 @@
                 )
               }
           )
-        end fingerprintSections
+        end sectionsByPotentialMatchKey
 
         val baseSectionsByFingerprint =
-          fingerprintSections(baseSources, pathInclusions.isIncludedOnBase)
+          sectionsByPotentialMatchKey(
+            baseSources,
+            pathInclusions.isIncludedOnBase
+          )
+        val basePotentialMatchKeys = baseSectionsByFingerprint.keySet
+          .asInstanceOf[collection.SortedSet[PotentialMatchKey]]
+
         val leftSectionsByFingerprint =
-          fingerprintSections(leftSources, pathInclusions.isIncludedOnLeft)
+          sectionsByPotentialMatchKey(
+            leftSources,
+            pathInclusions.isIncludedOnLeft
+          )
+        val leftPotentialMatchKeys = leftSectionsByFingerprint.keySet
+          .asInstanceOf[collection.SortedSet[PotentialMatchKey]]
+
         val rightSectionsByFingerprint =
-          fingerprintSections(rightSources, pathInclusions.isIncludedOnRight)
+          sectionsByPotentialMatchKey(
+            rightSources,
+            pathInclusions.isIncludedOnRight
+          )
+        val rightPotentialMatchKeys = rightSectionsByFingerprint.keySet
+          .asInstanceOf[collection.SortedSet[PotentialMatchKey]]
 
         @tailrec
-        def matchingFingerprintsAcrossSides(
-            baseFingerprints: Iterable[PotentialMatchKey],
-            leftFingerprints: Iterable[PotentialMatchKey],
-            rightFingerprints: Iterable[PotentialMatchKey],
+        // TODO: think of a better name...
+        def matchKeysAcrossSides(
+            basePotentialMatchKey: Option[PotentialMatchKey],
+            leftPotentialMatchKey: Option[PotentialMatchKey],
+            rightPotentialMatchKey: Option[PotentialMatchKey],
             matches: Set[GenericMatch],
             haveTrimmedMatches: Boolean
         ): MatchingResult =
-          // NOTE: when looking for matches, forbid any that would overlap or
-          // subsume an existing match at a larger or smaller window size. It is
+          // NOTE: when looking for matches, forbid any that would overlap an
+          // existing larger match or subsume an existing smaller match. It is
           // expected to find overlapping matches at the same window size,
           // especially when the optimal match size being searched for is
           // greater than `windowSize`.
           (
-            baseFingerprints.headOption,
-            leftFingerprints.headOption,
-            rightFingerprints.headOption
+            basePotentialMatchKey,
+            leftPotentialMatchKey,
+            rightPotentialMatchKey
           ) match
             case (Some(baseHead), Some(leftHead), Some(rightHead))
                 if potentialMatchKeyOrder.eqv(
@@ -1038,10 +1056,19 @@
                   )
                   matches -> true
 
-              matchingFingerprintsAcrossSides(
-                baseFingerprints.tail,
-                leftFingerprints.tail,
-                rightFingerprints.tail,
+              matchKeysAcrossSides(
+                basePotentialMatchKeys
+                  .iteratorFrom(baseHead)
+                  .drop(1)
+                  .nextOption(),
+                leftPotentialMatchKeys
+                  .iteratorFrom(leftHead)
+                  .drop(1)
+                  .nextOption(),
+                rightPotentialMatchKeys
+                  .iteratorFrom(rightHead)
+                  .drop(1)
+                  .nextOption(),
                 withAdditionalMatches,
                 haveTrimmedMatches || haveTrimmedAdditionalMatches
               )
@@ -1056,10 +1083,10 @@
                 ) =>
               // Tentatively synchronised the fingerprints between the base and
               // left, need to advance the right to resolve ...
-              matchingFingerprintsAcrossSides(
-                baseFingerprints,
-                leftFingerprints,
-                rightFingerprints.tail,
+              matchKeysAcrossSides(
+                basePotentialMatchKey,
+                leftPotentialMatchKey,
+                rightPotentialMatchKeys.iteratorFrom(baseHead).nextOption(),
                 matches,
                 haveTrimmedMatches
               )
@@ -1097,10 +1124,16 @@
                   )
                   matches -> true
 
-              matchingFingerprintsAcrossSides(
-                baseFingerprints.tail,
-                leftFingerprints.tail,
-                rightFingerprints,
+              matchKeysAcrossSides(
+                basePotentialMatchKeys
+                  .iteratorFrom(baseHead)
+                  .drop(1)
+                  .nextOption(),
+                leftPotentialMatchKeys
+                  .iteratorFrom(leftHead)
+                  .drop(1)
+                  .nextOption(),
+                rightPotentialMatchKey,
                 withAdditionalMatches,
                 haveTrimmedMatches || haveTrimmedAdditionalMatches
               )
@@ -1115,10 +1148,10 @@
                 ) =>
               // Tentatively synchronised the fingerprints between the base and
               // right, need to advance the left to resolve ...
-              matchingFingerprintsAcrossSides(
-                baseFingerprints,
-                leftFingerprints.tail,
-                rightFingerprints,
+              matchKeysAcrossSides(
+                basePotentialMatchKey,
+                leftPotentialMatchKeys.iteratorFrom(baseHead).nextOption(),
+                rightPotentialMatchKey,
                 matches,
                 haveTrimmedMatches
               )
@@ -1156,10 +1189,16 @@
                   )
                   matches -> true
 
-              matchingFingerprintsAcrossSides(
-                baseFingerprints.tail,
-                leftFingerprints,
-                rightFingerprints.tail,
+              matchKeysAcrossSides(
+                basePotentialMatchKeys
+                  .iteratorFrom(baseHead)
+                  .drop(1)
+                  .nextOption(),
+                leftPotentialMatchKey,
+                rightPotentialMatchKeys
+                  .iteratorFrom(rightHead)
+                  .drop(1)
+                  .nextOption(),
                 withAdditionalMatches,
                 haveTrimmedMatches || haveTrimmedAdditionalMatches
               )
@@ -1174,10 +1213,10 @@
                 ) =>
               // Tentatively synchronised the fingerprints between the left and
               // right, need to advance the base to resolve ...
-              matchingFingerprintsAcrossSides(
-                baseFingerprints.tail,
-                leftFingerprints,
-                rightFingerprints,
+              matchKeysAcrossSides(
+                basePotentialMatchKeys.iteratorFrom(leftHead).nextOption(),
+                leftPotentialMatchKey,
+                rightPotentialMatchKey,
                 matches,
                 haveTrimmedMatches
               )
@@ -1215,10 +1254,16 @@
                   )
                   matches -> true
 
-              matchingFingerprintsAcrossSides(
-                baseFingerprints,
-                leftFingerprints.tail,
-                rightFingerprints.tail,
+              matchKeysAcrossSides(
+                basePotentialMatchKey,
+                leftPotentialMatchKeys
+                  .iteratorFrom(leftHead)
+                  .drop(1)
+                  .nextOption(),
+                rightPotentialMatchKeys
+                  .iteratorFrom(rightHead)
+                  .drop(1)
+                  .nextOption(),
                 withAdditionalMatches,
                 haveTrimmedMatches || haveTrimmedAdditionalMatches
               )
@@ -1228,42 +1273,51 @@
               // minimum fingerprint to see if it can catch up and
               // synchronise...
 
-              val minimumFingerprint =
-                potentialMatchKeyOrder.min(
-                  baseHead,
-                  potentialMatchKeyOrder
-                    .min(leftHead, rightHead)
-                )
+              val baseLeftMinimum =
+                potentialMatchKeyOrder.min(baseHead, leftHead)
+              val baseRightMinimum =
+                potentialMatchKeyOrder.min(baseHead, rightHead)
+              val leftRightMinimum =
+                potentialMatchKeyOrder.min(leftHead, rightHead)
 
               // NOTE: just use `==` as we have already looked inside the
               // `PotentialMatchKey` instances - we just want to know which one
               // was the minimum.
-              if leftHead == minimumFingerprint
+              if baseLeftMinimum == leftRightMinimum
               then
-                matchingFingerprintsAcrossSides(
-                  baseFingerprints,
-                  leftFingerprints.tail,
-                  rightFingerprints,
+                // The left is the minimum.
+                matchKeysAcrossSides(
+                  basePotentialMatchKey,
+                  leftPotentialMatchKeys
+                    .iteratorFrom(baseRightMinimum)
+                    .nextOption(),
+                  rightPotentialMatchKey,
                   matches,
                   haveTrimmedMatches
                 )
               // NOTE: just use `==` as we have already looked inside the
               // `PotentialMatchKey` instances - we just want to know which one
               // was the minimum.
-              else if rightHead == minimumFingerprint
+              else if baseRightMinimum == leftRightMinimum
               then
-                matchingFingerprintsAcrossSides(
-                  baseFingerprints,
-                  leftFingerprints,
-                  rightFingerprints.tail,
+                // The right is the minimum.
+                matchKeysAcrossSides(
+                  basePotentialMatchKey,
+                  leftPotentialMatchKey,
+                  rightPotentialMatchKeys
+                    .iteratorFrom(baseLeftMinimum)
+                    .nextOption(),
                   matches,
                   haveTrimmedMatches
                 )
               else
-                matchingFingerprintsAcrossSides(
-                  baseFingerprints.tail,
-                  leftFingerprints,
-                  rightFingerprints,
+                // The base is the minimum.
+                matchKeysAcrossSides(
+                  basePotentialMatchKeys
+                    .iteratorFrom(leftRightMinimum)
+                    .nextOption(),
+                  leftPotentialMatchKey,
+                  rightPotentialMatchKey,
                   matches,
                   haveTrimmedMatches
                 )
@@ -1275,18 +1329,18 @@
               // synchronise...
               if potentialMatchKeyOrder.lt(baseHead, leftHead)
               then
-                matchingFingerprintsAcrossSides(
-                  baseFingerprints.tail,
-                  leftFingerprints,
-                  rightFingerprints,
+                matchKeysAcrossSides(
+                  basePotentialMatchKeys.iteratorFrom(leftHead).nextOption(),
+                  leftPotentialMatchKey,
+                  rightPotentialMatchKey,
                   matches,
                   haveTrimmedMatches
                 )
               else
-                matchingFingerprintsAcrossSides(
-                  baseFingerprints,
-                  leftFingerprints.tail,
-                  rightFingerprints,
+                matchKeysAcrossSides(
+                  basePotentialMatchKey,
+                  leftPotentialMatchKeys.iteratorFrom(baseHead).nextOption(),
+                  rightPotentialMatchKey,
                   matches,
                   haveTrimmedMatches
                 )
@@ -1297,18 +1351,18 @@
               // synchronise...
               if potentialMatchKeyOrder.lt(baseHead, rightHead)
               then
-                matchingFingerprintsAcrossSides(
-                  baseFingerprints.tail,
-                  leftFingerprints,
-                  rightFingerprints,
+                matchKeysAcrossSides(
+                  basePotentialMatchKeys.iteratorFrom(rightHead).nextOption(),
+                  leftPotentialMatchKey,
+                  rightPotentialMatchKey,
                   matches,
                   haveTrimmedMatches
                 )
               else
-                matchingFingerprintsAcrossSides(
-                  baseFingerprints,
-                  leftFingerprints,
-                  rightFingerprints.tail,
+                matchKeysAcrossSides(
+                  basePotentialMatchKey,
+                  leftPotentialMatchKey,
+                  rightPotentialMatchKeys.iteratorFrom(baseHead).nextOption(),
                   matches,
                   haveTrimmedMatches
                 )
@@ -1319,18 +1373,18 @@
               // synchronise...
               if potentialMatchKeyOrder.lt(leftHead, rightHead)
               then
-                matchingFingerprintsAcrossSides(
-                  baseFingerprints,
-                  leftFingerprints.tail,
-                  rightFingerprints,
+                matchKeysAcrossSides(
+                  basePotentialMatchKey,
+                  leftPotentialMatchKeys.iteratorFrom(rightHead).nextOption(),
+                  rightPotentialMatchKey,
                   matches,
                   haveTrimmedMatches
                 )
               else
-                matchingFingerprintsAcrossSides(
-                  baseFingerprints,
-                  leftFingerprints,
-                  rightFingerprints.tail,
+                matchKeysAcrossSides(
+                  basePotentialMatchKey,
+                  leftPotentialMatchKey,
+                  rightPotentialMatchKeys.iteratorFrom(leftHead).nextOption(),
                   matches,
                   haveTrimmedMatches
                 )
@@ -1340,12 +1394,15 @@
               // just a pair, so this terminates the recursion.
               withMatches(matches, windowSize, haveTrimmedMatches)
           end match
-        end matchingFingerprintsAcrossSides
+        end matchKeysAcrossSides
 
-        matchingFingerprintsAcrossSides(
-          baseSectionsByFingerprint.keySet,
-          leftSectionsByFingerprint.keySet,
-          rightSectionsByFingerprint.keySet,
+        // NASTY HACK: have to work around `SortedMultiDict` not exposing the
+        // more refined type of its key sets, it simply inherits the type
+        // signature from `MultiDict`.
+        matchKeysAcrossSides(
+          baseSectionsByFingerprint.keySet.headOption,
+          leftSectionsByFingerprint.keySet.headOption,
+          rightSectionsByFingerprint.keySet.headOption,
           matches = Set.empty,
           haveTrimmedMatches = false
         )
```

Still no improvement, seeing an execution time of 12 minutes.

Abandoning this approach, let's try set intersections next...

**Comment by sageserpent-open** [comment]
WIP on the set intersections approach in 5b18d7df73771ac25ec32083203d16b1a11bd5aa.

The good news is that for the most part, the tests pass, and the example runs in 6 minutes - so fat, so good, this is paying off.

The bad news is that there are occasional subtle differences in the matches found using the set approach versus the old linear scan approach - and these a) break `CodeMotionAnalysisExtensionTest.codeMotionAmbiguousWithAPreservation` b) cause some of the rejected test cases in `CodeMotionAnalysisExtensionTest.codeMotionAmbiguousWithAPreservation` to pass and c) cause final application error due to an overlapping section in the example:

```scala
com.sageserpent.kineticmerge.core.MappedContentSources$OverlappingSections: Overlapping section detected on side: BASE: 65fdb472 at path: /Users/gerardmurphy/IdeaProjects/kineticMerge-dogfood-examples/src/main/scala/com/sageserpent/kineticmerge/core/MoveDestinationsReport.scala, Section(
  label = "BASE: 65fdb472",
  path = /Users/gerardmurphy/IdeaProjects/kineticMerge-dogfood-examples/src/main/scala/com/sageserpent/kineticmerge/core/MoveDestinationsReport.scala,
  start = TextPosition(line = 31, characterOffset = 9),
  onePastEnd = TextPosition(line = 31, characterOffset = 36),
  startTokenIndex = 200,
  sizeInTokens = 4,
  content = "Left(leftElement)          "
) (content: Vector(Significant(Left), Significant((), Significant(leftElement), WithTrailingWhitespace(Significant()),Whitespace(          )))) overlaps with start of section: Section(
  label = "BASE: 65fdb472",
  path = /Users/gerardmurphy/IdeaProjects/kineticMerge-dogfood-examples/src/main/scala/com/sageserpent/kineticmerge/core/MoveDestinationsReport.scala,
  start = TextPosition(line = 31, characterOffset = 14),
  onePastEnd = TextPosition(line = 31, characterOffset = 39),
  startTokenIndex = 202,
  sizeInTokens = 4,
  content = "leftElement)          => "
) (content: Vector(Significant(leftElement), WithTrailingWhitespace(Significant()),Whitespace(          )), Significant(=), WithTrailingWhitespace(Significant(>),Whitespace( )))), overlap content: Vector(Significant(leftElement), WithTrailingWhitespace(Significant()),Whitespace(          ))).
```

Hmm...

**Comment by sageserpent-open** [comment]
More good news, and more bad news...

The good:

As of commit 55ae7adfb5cde922701335d6a17225548ec2850e, `CodeMotionAnalysisExtensionTest.codeMotionAmbiguousWithAPreservation` passes and the rejected test cases in `CodeMotionAnalysisExtensionTest.codeMotionAmbiguousWithAPreservation` are consistent with the behaviour seen prior to using the set intersection approach.

The example merge succeeds again, and the new implementation is lot less sprawling than the old one - `matchKeysAcrossSides` is a lot easier to understand.

Running `CodeMotionAnalysisExtensionTest.matchingSectionsAreFound` yields the following number of test cases in 5 minutes:

Commit c9b01d5036d4029395bc4b57d899e743339e35fe (linear scan approach): 1007, 1031

Commit 55ae7adfb5cde922701335d6a17225548ec2850e (set intersections approach): 1297, 1191, 1268

So a modest improvement in throughput in the tests.

That bad:

The example merge took 12 minutes to run!

Hot spot report:
<img width="1745" alt="Screenshot 2024-12-17 at 11 43 43" src="https://github.com/user-attachments/assets/783bad24-04ee-4af6-8f46-2fce47cb5129" />

**Comment by sageserpent-open** [comment]
Now, the hashing of the `PotentialMatchKey` instances has been pushed right down, so something is going well here. Fingertree is dominating the self-time, so it's the obvioust next candidate to look at.

The fingerprinting is starting to look significant too, so that's something to think about in the longer term...

**Comment by sageserpent-open** [comment]
One last try with the set intersection approach, as of 067f737374c61c2cd643d07e93cccdb4c72bb9f9, got the example merge to 9 minutes.

Looking at these now, using the linear scan approach...

> 3\. All-sides matches block the presence of smaller matches whose sections are subsumed on three sides of the larger all-sides match. This is done using fingertrees for each side, but we can look for an earlier fast-path by blocking out the content regions covered by the matched sections from the initial search for fingerprints - there is no point in looking for smaller matches whose sections would lie within these blocked out regions. I'm not 100% sure about this, but it's worth exploring.

> 4\. Thinking about optimising the tests for section overlap with, subsumption of or being subsumed by a previously matched section, it is worth extracting a separate strategy API for this, writing some tests for it using the existing fingertree as an initial SUT, then cutting over to use an interval tree or whatever. Cats has a DIET tree that might fit that bill, as might `RangeOfSlots`.

**Comment by sageserpent-open** [comment]
Added support for fingerprinted inclusions with knockouts in 59b4157 - this is to address:

> 3\. All-sides matches block the presence of smaller matches whose sections are subsumed on three sides of the larger all-sides match. This is done using fingertrees for each side, but we can look for an earlier fast-path by blocking out the content regions covered by the matched sections from the initial search for fingerprints - there is no point in looking for smaller matches whose sections would lie within these blocked out regions. I'm not 100% sure about this, but it's worth exploring.

Running CodeMotionAnalysisExtensionTest.matchingSectionsAreFound yields the following number of test cases in 5 minutes: 1105, 1190, 1365.

Running the example merge using a build from this commit takes - wait for it - 1 minute and 44 seconds!

To sanity check, running the example merge using a build from `main` at cd24ef007ff8e9c6499f1ec4269633b323299e0a takes a whopping 47 minutes and 43 seconds. This is consistent with what this ticket started with:
> Overall time from Visual VM: 46 minutes, 49 minutes, 48 minutes.

The resulting merges from the build off 59b4157 and off cd24ef007ff8e9c6499f1ec4269633b323299e0a are identical, so we've not lost fidelity.

Now, that's a result.

**Comment by sageserpent-open** [comment]
Commit 9f4b6bfdb3a8a425daea4f86b068fe72f9942737 adds a finer-grained parallelisation of the fingerprinted inclusions, which approximates the goal of this idea from above:

> This could be extended so that for large files, fingerprinting takes place in parallel swathes. Although this means that each swathe needs to be primed beforehand (which is optimised away when the entire file is fingerprinted in one swathe), for large files one hopes that being able to parallelize outweighs this overhead.

Using this approach means that fingerprints for large window sizes aren't parallelised that much (if at all), but that is probably a good thing - if the window size is of a similar order of magnitude as the target file size, most of the overhead goes into priming the first fingerprint, so it is better to stick with just that and one swathe of fingerprints off it.

As it turns out, this makes little difference, the example merge takes 1 minue 34 seconds, 1 minute 37 seconds, but there's no harm in leaving it in as a speculative optimisation for general merging.

**Comment by sageserpent-open** [comment]
At this stage, the running time is down to the point where the file-by-file merge algorithm starts to become significant, of the order of 10 seconds is taken by that.

**Comment by sageserpent-open** [comment]
Merged on the work on set intersection with the fingerprinted inclusion work, this gives commit d6b2da9487c16ca46d4cc1ea469dd094afd7c03d.

Running CodeMotionAnalysisExtensionTest.matchingSectionsAreFound yields the following number of test cases in 5 minutes: 1808, 1638.

The example merge runs in:

a) 1 minute 17 seconds, of which manual stopwatch timing had the final post-matching phase taking roughly 20 seconds.
b) 1 minute 19 seconds, , of which manual stopwatch timing had the final post-matching phase taking roughly 20 seconds.

**Comment by sageserpent-open** [comment]
Experimenting with how much parallelisation to employ...

Commit a531821e4113b64b68a53b0303293471806aeb1e, essentially this is d6b2da9 with dependency upgrades, including the Scala Parallel Collections. It parallelises on two levels, up by path and down by content slice.

Execution times for example merge:

1 minute 20 seconds (4433)
1 minute 20 seconds (4062)
1 minute 21 seconds (3688)
1 minute 21 seconds (3317)

Commit 48c967e4127e7a0d4034344e8e97c8139c565e58, this flattens the sequences of contents slices over all paths beforehand, doing parallelisation at just one level.

Execution times for example merge:

1 minute 23 seconds (1538)
1 minute 20 seconds (1931)
1 minute 22 seconds (2302)
1 minute 21 seconds (2680)

There isn't much difference, so on that basis I prefer just on elevel of parallelisation in 48c967e4127e7a0d4034344e8e97c8139c565e58.

**Comment by sageserpent-open** [comment]
The hot spot reports for 48c967e vary considerably, there doesn't appear to be any clear candidate for further optimisation.

Sometimes in the fingertree stuff is towards the top, sometimes its the fingerprinting stuff. There is no clear pattern from one run to another.

The only thing that does reliably dominate are calls to `Statics.anyHash`, which are from all over the place, as this is a low-level part of the Scala runtime library.

The hot spot reports for a531821 exhibit the same behaviour.

What about the work done on fingerprinting in swathes? Would that help?

**Comment by sageserpent-open** [comment]
|Maximum swathe size|Execution time (seconds)|
|---|---|
|1000|78|
|2000|79|
|5000|79
|500|79|
|200|78|
|100|80|
|10000|79|
|15000|78|

This is from commit: 862aabdfec1673a718b637089ad5392a4b5f69ac

It's a nice idea, but it makes no difference whatsoever.

**Comment by sageserpent-open** [comment]
I think that's a wrap now - time to get a release out to commemorate the end of 2024.

One last thing to do - the final per-file merges take around 20 seconds for the example merge. That isn't really that long, given how pathological the example is, but it is disconcerting to see the application just sit there apparently doing nothing after seeing the matching progress bar steadily counting down.

Let's add a progress bar to the per-file merge...

**Comment by sageserpent-open** [comment]
No, adding a progress bar to the per-file merge (actually it goes down the longest common subsequence implementation) seems to result in a strange backwards-and-forwards progress bar, plus a lot separate progress bars for each little merge, which ends up filling the console output with mysterious blank lines once the progress bars are erased.

If this becomes a problem for folk in practice, then it's worth revisiting in a separate ticket.

**Comment by sageserpent-open** [comment]
Branch `main` fast-forwarded to 8464ce39d29573866fce8e26f98f96119a3b4de7.

**Comment by sageserpent-open** [comment]
This went out in Release 1.4.0.
